### PR TITLE
interpreter: extract_objects provides a valid source

### PR DESCRIPTION
**Pull request by bonzini**
This ensures that there is no warnings when running meson on `test cases/common/22 object extraction`.

### CODE REVIEW

**Comment by bonzini** [comment]
Ping (for 0.60.1 even?)

File: mesonbuild/modules/unstable_rust.py [hunk 1]
```diff
@@ -203,6 +203,8 @@ def bindgen(self, state: 'ModuleState', args: T.List, kwargs: 'FuncBindgen') ->
         name: str
         if isinstance(header, File):
             name = header.fname
+        elif isinstance(header, (BuildTarget, BothLibraries, ExtractedObjects)):
+            raise InterpreterException('bindgen source file must be a C header, not an object or build target')
```
Comment (nirbheek) [inline mesonbuild/modules/unstable_rust.py:207]: Why was this added? It seems unsafe to put this into a stable release?

Comment (nirbheek) [reply mesonbuild/modules/unstable_rust.py:207]: Actually, this whole commit seems to be unrelated to the PR?

Comment (bonzini) [reply mesonbuild/modules/unstable_rust.py:207]: It's in the commit message (not your fault, GitHub review is horrible):

> This also fixes a mypy failure from the next patch, since ExtractedObjects does not have a zero-argument get_outputs method.

Comment (nirbheek) [reply mesonbuild/modules/unstable_rust.py:207]: Okay. So I'm wondering, is it possible that someone is doing this nonsensical thing in their build files and it accidentally works somehow right now, and it will break if this goes into a stable release?

Comment (bonzini) [reply mesonbuild/modules/unstable_rust.py:207]: In other words, ExtractedObjects as in the new testcase would have caused a Python backtrace, while BuildTarget and BothLibraries simply do not make sense in this position (I tried quickly `bindgen /bin/ls` and bindgen panics very soon).

Comment (nirbheek) [reply mesonbuild/modules/unstable_rust.py:207]: Okay, thanks :)

**Comment by nirbheek** [comment]
LGTM
